File: Sorting Algorithm/Bubble Sort/BubbleSort.py
#=============================================================================================================================
#      Project:                  Bubble Sort
#      Time Complexity:          O(n2)
#      Space Complexity:         O(1)
#      Stability:                Stable
#=============================================================================================================================

# Basic Bubble Sort
def BubbleSort(L):
	for unsorted in range(len(L)-1, 0, -1):
		for i in range(unsorted):
			if L[i] > L[i+1]:
				L[i], L[i+1] = L[i+1], L[i]


# Bubble Sort with flag to indicate the index after which the array has been sorted
def BubbleSort2(L):
	flag = len(L)-1
	while flag > 0:
		origin_flag = flag
		for i in range(0,origin_flag):
			if L[i] > L[i+1]:
				flag = i
				L[i], L[i+1] = L[i+1], L[i]
		if flag == origin_flag:
			flag = 0



L1 = [2, 3, 5, 1, 7, 6, 8, 9, 10, 4]
L2 = [2, 3, 5, 1, 7, 6, 8, 9, 10, 4]

BubbleSort(L1)
print "BubbleSort: ", L1

BubbleSort2(L2)
print "BubbleSort2: ", L2

File: Sorting Algorithm/Bubble Sort/BubbleSort.cpp
//=============================================================================================================================
//      Project:                  Bubble Sort
//      Time Complexity:          O(n2)
//      Space Complexity:         O(1)
//      Stability:                Stable
//=============================================================================================================================

#ifndef bubbleSort
#define bubbleSort

// Basic Bubble Sort
void BubbleSort(int a[], int n)
{
	int temp;
	for (int i = 0; i < n-1; ++i)
	{
		for (int j = 0; j < n-i-1; ++j)
		{
			if (a[j] > a[j+1])
			{
				temp = a[j];
				a[j] = a[j+1];
				a[j+1] = temp;
			}
		}
	}
}

// Bubble Sort with flag to indicate the index after which the array has been sorted
void BubbleSort2(int a[], int n)
{
	int temp;
	int flag = n-1;     // elements after flag should be sorted, at the begining, we assume all elements all unsorted.
	while (flag > 0)    // if flag = 0, the whole array has been sorted
	{
		int origin_flag = flag;       

		for (int i = 0; i < origin_flag; ++i)
		{
			if (a[i] > a[i+1])
			{
				flag = i;
				temp = a[i];
				a[i] = a[i+1];
				a[i+1] = temp;
			}
		}

		if ( flag == origin_flag)
		{
			flag = 0;
		}
	}
}

#endif